### PR TITLE
Case 22379: Run-time switch for deferred/forward rendering

### PR DESCRIPTION
**Pull request by SamGondelman**
https://highfidelity.manuscript.com/f/cases/22379/Run-time-switch-for-Deferred-Forward-rendering

Test plan:
- Rendering should look the same as master on Quest and Desktop.
- On Desktop (only), you can switch between Forward and Deferred rendering by running this in the console:
    - Forward: `Render.getConfig("RenderMainView.DeferredForwardSwitch").branch = 1`
    - Deferred: `Render.getConfig("RenderMainView.DeferredForwardSwitch").branch = 0`
- In Deferred mode, you'll see things like shadows, haze, and local lights.  In Forward mode, you won't see those things and might see other rendering differences.
- In deferred mode, run [this](https://gist.githubusercontent.com/SamGondelman/1b7418c416df22ddc84683b72e23a2fd/raw/fc5f092e8ca728e2c7a78c73c751d6f936c13b58/RenderLayer.js).  You'll see something like this:
image
- Switch to forward mode and nothing (/very little) should change.
- Press "u" to make all of the entities (except the model, polyline, and material) slightly transparent.  They should still look the same in forward/deferred.
- Press "o" to cycle through the 3 render layers.  In the last one, the HUD layer, everything will appear slightly different (color grading and AA), but they should look the same between forward and deferred.
- Known issues (on master already):
    - the polyvox entity doesn't seem to render
    - when you toggle the transparency, the particle gets a speedSpread of 1?
- The blue and orange teleport parabolas should render in both deferred and forward modes.

### CODE REVIEW

File: libraries/task/src/task/Config.h [hunk 1]
```diff
@@ -262,6 +262,18 @@ public slots:
     void refresh();
 };
 
+class SwitchConfig : public TaskConfig {
+    Q_OBJECT
+    Q_PROPERTY(bool switchIndex READ getSwitchIndex WRITE setSwitchIndex NOTIFY dirtyEnabled)
```
Comment (SamGondelman) [inline libraries/task/src/task/Config.h]: Open to suggestions for a better name than "switchIndex"

File: libraries/task/src/task/Task.h [hunk 2]
```diff
@@ -244,25 +247,6 @@ class Task : public Job<JC, TP> {
...
-        typename Jobs::const_iterator getJob(std::string name) const {
-            typename Jobs::const_iterator jobIt;
-            for (jobIt = _jobs.begin(); jobIt != _jobs.end(); ++jobIt) {
-                if (jobIt->getName() == name) {
-                    return jobIt;
-                }
-            }
-            return jobIt;
-        }
-
```
Comment (SamGondelman) [inline libraries/task/src/task/Task.h]: unused

File: libraries/task/src/task/Task.h [hunk 3]
```diff
@@ -357,6 +341,76 @@ class Task : public Job<JC, TP> {
     template <class T, class O, class C = Config> using ModelO = TaskModel<T, C, None, O>;
     template <class T, class I, class O, class C = Config> using ModelIO = TaskModel<T, C, I, O>;
 
+    template <class T, class C = SwitchConfig, class I = None, class O = None> class SwitchTaskModel : public TaskModel<T, C, I, O> {
```
Comment (SamGondelman) [inline libraries/task/src/task/Task.h]: template nonsense but this is basically a wrapper around a TaskModel that runs all the _jobs and the picks the branching jobs to run based on switchIndex

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/19034/HighFidelity-Beta-PR15469-fc2c0bb.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/19034/HighFidelity-Beta-Interface-PR15469-fc2c0bb.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/19036/HighFidelity-Beta-PR15469-ca5c7e3.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/19036/HighFidelity-Beta-Interface-PR15469-ca5c7e3.dmg

**Comment by hifi-gustavo** [comment]
Android build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/19036/HighFidelity-Beta-PR15469-ca5c7e3.apk).  Quest build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/19036/HighFidelity-Quest-Beta-PR15469-ca5c7e3.apk)

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/19036/HighFidelity-Beta-PR15469-ca5c7e3.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/19036/HighFidelity-Beta-Interface-PR15469-ca5c7e3.exe

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/19080/HighFidelity-Beta-PR15469-4a02ab4.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/19080/HighFidelity-Beta-Interface-PR15469-4a02ab4.exe

**Comment by SamGondelman** [comment]
build this please

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/19086/HighFidelity-Beta-PR15469-4a02ab4.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/19086/HighFidelity-Beta-Interface-PR15469-4a02ab4.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/19091/HighFidelity-Beta-PR15469-674ad50.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/19091/HighFidelity-Beta-Interface-PR15469-674ad50.dmg

**Comment by hifi-gustavo** [comment]
Android build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/19091/HighFidelity-Beta-PR15469-674ad50.apk).  Quest build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/19091/HighFidelity-Quest-Beta-PR15469-674ad50.apk)

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/19091/HighFidelity-Beta-PR15469-674ad50.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/19091/HighFidelity-Beta-Interface-PR15469-674ad50.exe

File: libraries/render-utils/src/RenderViewTask.cpp [hunk 4]
```diff
@@ -10,32 +10,46 @@
...
 
-    if (isDeferred) {
-        // Warning : the cull functor passed to the shadow pass should only be testing for LOD culling. If frustum culling
-        // is performed, then casters not in the view frustum will be removed, which is not what we wish.
-        const auto shadowTaskIn = RenderShadowTask::Input(lightingStageFramesAndZones.get<AssembleLightingStageTask::Output>().get0()[0], lightingModel).asVarying();
-        const auto shadowTaskOut = task.addJob<RenderShadowTask>("RenderShadowTask", shadowTaskIn, cullFunctor, tagBits, tagMask);
-
-        const auto renderInput = RenderDeferredTask::Input(items, lightingModel, lightingStageFramesAndZones, shadowTaskOut).asVarying();
-        task.addJob<RenderDeferredTask>("RenderDeferredTask", renderInput);
+    if (!DISABLE_DEFERRED) {
```
Comment (samcake) [inline libraries/render-utils/src/RenderViewTask.cpp]: I would suggest to replace this if block by a #ifdef NOT Q_OS_ANDROID
ANd use the info of DISABLE_DEFERRED to configure the default value of the switch branch with it

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/19330/HighFidelity-Beta-PR15469-095876e.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/19330/HighFidelity-Beta-Interface-PR15469-095876e.exe

File: libraries/render-utils/src/GeometryCache.cpp [hunk 5]
```diff
@@ -2162,6 +2158,7 @@ void GeometryCache::bindWebBrowserProgram(gpu::Batch& batch, bool transparent) {
 gpu::PipelinePointer GeometryCache::getWebBrowserProgram(bool transparent) {
     static std::once_flag once;
     std::call_once(once, [&]() {
+        // FIXME: need a forward pipeline for this
```
Comment (SamGondelman) [inline libraries/render-utils/src/GeometryCache.cpp]: The FIXMEs sprinkled throughout this PR aren't immediately necessary.

File: libraries/render-utils/src/RenderViewTask.cpp [hunk 6]
```diff
@@ -10,32 +10,49 @@
...
+    // is performed, then casters not in the view frustum will be removed, which is not what we wish.
+    const auto shadowTaskIn = RenderShadowTask::Input(lightingStageFramesAndZones.get<AssembleLightingStageTask::Output>().get0()[0], lightingModel).asVarying();
+    const auto shadowTaskOut = task.addJob<RenderShadowTask>("RenderShadowTask", shadowTaskIn, cullFunctor, tagBits, tagMask);
+
+    const auto renderDeferredInput = RenderDeferredTask::Input(items, lightingModel, lightingStageFramesAndZones, shadowTaskOut).asVarying();
+    task.addJob<RenderDeferredTask>("RenderDeferredTask", renderDeferredInput);
+}
+
+void DeferredForwardSwitchJob::build(JobModel& task, const render::Varying& input, render::Varying& output, render::CullFunctor cullFunctor, uint8_t tagBits, uint8_t tagMask) {
+    task.addBranch<RenderShadowsAndDeferredTask>("RenderShadowsAndDeferredTask", RENDER_FORWARD ? 1 : 0, input, cullFunctor, tagBits, tagMask);
```
Comment (SamGondelman) [inline libraries/render-utils/src/RenderViewTask.cpp]: @samcake I'm not sure if this is what you wanted.  If `RENDER_FORWARD`, I actually swap the two branches.  Alternatively, I could change the default branch, but I'm not sure where best to do that?  Via the config?  Can I do that in `build`?

File: libraries/render/src/render/Args.h [hunk 7]
```diff
@@ -64,6 +64,7 @@ namespace render {
     public:
         enum RenderMode { DEFAULT_RENDER_MODE, SHADOW_RENDER_MODE, MIRROR_RENDER_MODE, SECONDARY_CAMERA_RENDER_MODE };
         enum DisplayMode { MONO, STEREO_MONITOR, STEREO_HMD };
+        enum RenderMethod { DEFERRED, FORWARD };
```
Comment (SamGondelman) [inline libraries/render/src/render/Args.h]: Is there a better name for this?

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/19327/HighFidelity-Beta-PR15469-c529ecb.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/19327/HighFidelity-Beta-Interface-PR15469-c529ecb.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/19334/HighFidelity-Beta-PR15469-2cedc76.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/19334/HighFidelity-Beta-Interface-PR15469-2cedc76.dmg

**Comment by hifi-gustavo** [comment]
Android build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/19334/HighFidelity-Beta-PR15469-2cedc76.apk).  Quest build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/19334/HighFidelity-Quest-Beta-PR15469-2cedc76.apk)

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/19336/HighFidelity-Beta-PR15469-f1d7e75.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/19336/HighFidelity-Beta-Interface-PR15469-f1d7e75.exe

**Comment by hifi-gustavo** [comment]
Android build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/19336/HighFidelity-Beta-PR15469-f1d7e75.apk).  Quest build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/19336/HighFidelity-Quest-Beta-PR15469-f1d7e75.apk)

**Comment by SamGondelman** [comment]
build this please

**Comment by hifi-gustavo** [comment]
Android build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/19342/HighFidelity-Beta-PR15469-f1d7e75.apk).  Quest build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/19342/HighFidelity-Quest-Beta-PR15469-f1d7e75.apk)

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/19342/HighFidelity-Beta-PR15469-f1d7e75.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/19342/HighFidelity-Beta-Interface-PR15469-f1d7e75.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/19342/HighFidelity-Beta-PR15469-f1d7e75.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/19342/HighFidelity-Beta-Interface-PR15469-f1d7e75.exe

File: libraries/task/src/task/Task.h [hunk 8]
```diff
@@ -369,8 +353,150 @@ class Task : public Job<JC, TP> {
...
+            }
+        }
+
+        void run(const ContextPointer& jobContext) override {
+            auto config = std::static_pointer_cast<C>(Concept::_config);
+            if (config->isEnabled()) {
+                auto jobsIt = SwitchConcept::_branches.find(config->getBranch());
+                if (jobsIt != SwitchConcept::_branches.end()) {
+                    jobsIt->second.run(jobContext);
+                    if (jobContext->taskFlow.doAbortTask()) {
```
Comment (samcake) [inline libraries/task/src/task/Task.h]: As we discussed let s remove that from here, we don;t need to check this here

File: libraries/render/src/render/Args.h (hunk 7 above)
Comment (samcake) [reply libraries/render/src/render/Args.h]: THis is good enough

**Comment by hifi-gustavo** [comment]
Android build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/19367/HighFidelity-Beta-PR15469-eb8c722.apk).  Quest build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/19367/HighFidelity-Quest-Beta-PR15469-eb8c722.apk)

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/19367/HighFidelity-Beta-PR15469-eb8c722.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/19367/HighFidelity-Beta-Interface-PR15469-eb8c722.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/19367/HighFidelity-Beta-PR15469-eb8c722.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/19367/HighFidelity-Beta-Interface-PR15469-eb8c722.exe

**Comment by hifi-gustavo** [comment]
Android build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/19374/HighFidelity-Beta-PR15469-eb8c722.apk).  Quest build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/19374/HighFidelity-Quest-Beta-PR15469-eb8c722.apk)

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/19374/HighFidelity-Beta-PR15469-eb8c722.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/19374/HighFidelity-Beta-Interface-PR15469-eb8c722.exe

**Pull request merged**